CHORE: Added model and field valdators, customised admin dashboard

thenoblet opened the pull request:
- Created `validators.py` to centralise custom validation logic.
- Implemented model field validation for `User`, `UserProfile`, `Organisation` and `Role` models.
- Modified `models.py` to integrate validation with `clean()` methods.
- Customised the admin dashboard in `admin.py` for improved usability.
- Added a migration (`0002_alter_organisation_table.py`) to update the `Organisation` table schema.

These changes enhance data integrity and improve the user experience in the admin interface.

thenoblet committed:
CHORE: Added model and field valdators, customised admin dashboard

- Created `validators.py` to centralise custom validation logic.
- Implemented model field validation for `User`, `UserProfile`, `Organisation` and `Role` models.
- Modified `models.py` to integrate validation with `clean()` methods.
- Customised the admin dashboard in `admin.py` for improved usability.
- Added a migration (`0002_alter_organisation_table.py`) to update the `Organisation` table schema.

nanafox commented on src/user_service/models.py:
```diff
@@ -248,14 +292,15 @@ class Organisation(models.Model):
         """Metadata for the organisations model."""
 
-        db_table = "organisations"
+        db_table = "organisation"
```
Shouldn't table names be plural?

thenoblet replied:
Anything you say boss

nanafox commented on src/user_service/validators.py:
```diff
@@ -0,0 +1,309 @@
+
+        try:
+            owner_field = Organisation._meta.get_field("owner")
+        except FieldDoesNotExist:
```
When you catch an exception and raise another one, you should be explicit about it. You can for example do something like
```python
try:
    something_that_may_raise_an_exception
except ThOriginalException as e:
    raise SomeOtherExceptionInstead from e

thenoblet replied:
Well, exception chaining is good. Why not? Will help with logging... we can see the full chain of exceptions in the logs when debugging. 

I will make the change

nanafox commented on src/user_service/models.py:
```diff
@@ -157,31 +166,56 @@ class UserStation(models.TextChoices):
     )
     organisation = models.OneToOneField(
-        "Organisation", on_delete=models.CASCADE, null=True, blank=True
+        "Organisation", on_delete=models.CASCADE, null=True, blank=True, related_name="user_org"
```
The default related name Django gives is `organisation` which I think is enough. This makes it simpler to have something like
```python
user = User.objects.get(id=1)
user.organization  # insetad of having to do user.user_org
```

What do you think?

thenoblet replied:
You'll have a problem with me on this project.... I write BRE you write AME. Massa

nanafox reviewed (changes requested):
I've checked the additions and for the most part, it is good. Before we proceed to merge, how have they been tested to confirm they work? Do you have any screenshots you can share to verify that the validations and everything we have here works as intended?

thenoblet commented:
> I've checked the additions and for the most part, it is good. Before we proceed to merge, how have they been tested to confirm they work? Do you have any screenshots you can share to verify that the validations and everything we have here works as intended?

Yep. Tested and works.

thenoblet committed:
CHANGES: Addressed suggested changes

- Model validators has been refactored in accordance with the suggested changes
